### ISSUE TITLE
Hardware PMU - performance counters using Linux perf and /resctrl interface

### ISSUE THREAD

**Issue by ss7pro** [google/cadvisor#2388]
Hey,

In Intel we're working on a solution that is providing online workload characterization using hardware PMU using Linux perf interface. In addition we're also using RDT through /resctrl interface to extract memory bandwidth and cache usage. Full list of metrics that we collect is documented here:

https://github.com/intel/workload-collocation-agent/blob/master/docs/metrics.rst

We would like to migrate all telemetry collection and query to cAdvisor. Do you think it make sense to add those kind of metrics to cAdvisor and if yes do you anticipate any major issues in that approach?

**Comment by dashpole** [comment]
cAdvisor exposes metrics associated with cgroups.  From looking at you metric list, it looks like most metrics are associated with tasks, although looking through some of the other docs, it looks like "task" can at least sometimes mean cgroup.  Are all "task" metrics associated with cgroups?

**Comment by iwankgb** [comment]
@dashpole some of the task metrics come from resctrl filesystem (they are related to Intel RDT - cache and memory bandwidth monitoring). Task concept is similar to pod in Kubernetes: it always is a cgroup, it might include other cgroups and it might rely on other kernel interfaces to collect metrics too.

**Comment by dashpole** [comment]
Given a particular cgroup, is it possible to get the associated task?

Are the kernel interfaces it relies on documented anywhere?

**Comment by iwankgb** [comment]
Resctrl is [documented in kernel](https://www.kernel.org/doc/html/latest/x86/resctrl_ui.html). 

Regarding identifying a task by cgroup - I'm not sure if I understand your question; are you asking about implementation of a task in Workload Collocation Agent?

**Comment by iwankgb** [comment]
BTW, as far as I can tell RDT and resctrl are already being handled using runc's `IntelRdtManager`, so we will be able to utilize [existing functionality](https://github.com/google/cadvisor/blob/master/vendor/github.com/opencontainers/runc/libcontainer/container_linux.go#L183
), I hope.

**Comment by dashpole** [comment]
Oh, great.  I see it is already integrated into libcontainer.  In that case, it will be easy to integrate.

**Comment by iwankgb** [comment]
cAdvisor is using libcontainer to collect all the metrics from cgroups - am I right? If this is the case then we will proceed with some PRs in libcontainer and then we will get back to you once we are ready to integrate them to cAdvisor.

**Comment by dashpole** [comment]
Yes.  cAdvisor uses libcontainer to collect metrics from cgroups.

**Comment by iwankgb** [comment]
@dashpole I think that you can see full list of TODOs above.

**Comment by bboreham** [comment]
> Full list of metrics that we collect is documented here:
> https://github.com/intel/workload-collocation-agent/blob/master/docs/metrics.rst

This link 404s today; is there a new location that lists the metrics?

**Comment by Creatone** [comment]
@bboreham Unfortunately the repository has been archived. However, here is my fork: https://github.com/Creatone/workload-collocation-agent/blob/master/docs/metrics.rst

**Comment by dims** [comment]
Closing this issue as it has been open since before January 1, 2024. If this is still relevant, please feel free to reopen or create a new issue.
